File: algorithms/DP/L198_House_Robber.cpp
/**
 * Leetcode 198) https://leetcode.com/problems/house-robber/
 *
 * House Robber
 */

#include <iostream>
#include <vector>

using namespace std;

/* ----- Solution ----- */
class Solution {
    int N;
    vector<int> dp;
    int robImpl(const vector<int>& nums, const int& start) {
        if(start == N - 1) {
            return nums[start];
        } else if(start >= N) {
            return 0;
        } else if(dp[start] != INT_MIN) {
            return dp[start];
        }
        return dp[start] = max(robImpl(nums, start + 2) + nums[start], robImpl(nums, start + 3) + nums[start + 1]);
    }
public:
    int rob(vector<int>& nums) {
        N = nums.size();
        dp.clear();
        dp.resize(N, INT_MIN);
        return robImpl(nums, 0);
    }
};

/* ----- Testing ----- */
struct Testing {
    vector<int> nums;
    int outp;
    Testing(): Testing({}, 0) {}
    Testing(vector<int> n, int o): nums(n), outp(o) {}
    static const int SIZE = 3;
};

int main() {
    Testing t[Testing::SIZE] = {
        Testing({1,2,3,1}, 4),
        Testing({2,7,9,3,1}, 12),
    };

    Solution s;
    for(auto& tcase : t) {
        auto res = s.rob(tcase.nums);
        if(res != tcase.outp) {
            cout << "FAIL: want " << tcase.outp << " got " << res << endl;
            return 1;
        }
    }

    cout << "ok" << endl;
    return 0;
}

File: problem-archive/DP/L198_House_Robber.java
package com.java.app;

class Solution {
    public int rob(int[] nums) {
        int len = nums.length;
        int max = -2147483648;
        for (int i = 0; i < len; i++) {
            int b = i > 0 ? nums[i - 1] : 0;
            int bb = i > 1 ? nums[i - 2] : 0;
            nums[i] = Math.max(b, bb + nums[i]);
            max = Math.max(max, nums[i]);
        }
        return max;
    }
}

class Testing {
    public int[] nums;
    private int output;

    public Testing(int[] nums, int output) {
        this.nums = nums;
        this.output = output;
    }

    public boolean test(int output) {
        return this.output == output;
    }
}

public class L198_House_Robber {
    public static void main(String[] args) {
        int[][] numses = {
            {1,2,3,1},
            {2,7,9,3,1},
            {2,1},
            {1},
            {2,1,1,2},
            {1,3,1,3,100},
        };

        Testing[] tcases = {
            new Testing(numses[0], 4),
            new Testing(numses[1], 12),
            new Testing(numses[2], 2),
            new Testing(numses[3], 1),
            new Testing(numses[4], 4),
            new Testing(numses[5], 103),
        };

        Solution s = new Solution();

        for (Testing t : tcases) {
            if (!t.test(s.rob(t.nums))) {
                System.err.println("FAIL");
                return;
            }
        }

        System.out.println("ok");
    }
}
